src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

```diff
@@ -466,10 +467,34 @@ public HttpClient CreateDefaultClient(params DelegatingHandler[] handlers)
         {
             EnsureServer();
 
+            (HttpMessageHandler, Uri?) CreateHandler()
+            {
+                if (_server is null)
+                {
+                    throw new InvalidOperationException("Server not available");
+                }
+
+                if (_server is TestServer testServer)
+                {
+                    return (testServer.CreateHandler(), null);
+                }
+
+                var httpAddress = _server.Features.Get<IServerAddressesFeature>()?.Addresses.FirstOrDefault(a => a.StartsWith("http://"));
```

Comment:
If there isn’t an HTTP address, try looking for an HTTPS one before throwing?

For browser UI testing I usually configure Kestrel with a self-signed test cert so that I don’t need to disable RequireHttps for MVC or worry about behaviours in auth that require HTTPS.

Example: https://github.com/martincostello/dotnet-minimal-api-integration-testing/blob/1493d95e9bb4458ac8d9b89f414e1396a55c8804/tests/TodoApp.Tests/HttpServerFixture.cs#L70-L76

Reply:
Yes, I should do that.

Reply:
Yes I will do that. Are you calling EnsureServer manually btw?

Reply:
I don't use the built-in one (at least not explicitly), instead I manually construct my own server as part of the xunit fixture init: https://github.com/martincostello/dotnet-minimal-api-integration-testing/blob/1493d95e9bb4458ac8d9b89f414e1396a55c8804/tests/TodoApp.Tests/HttpServerFixture.cs#L25

Reply:
To make this cleaner you need to: 
- Have the ability to start the server (I'd prefer if you didn't need to manually do it)
- Get a handle on the address so you can construct a client

PS: WebApplicationFactory feels super messy because it has so many virtual methods which make it hard to modify 😄

Reply:
Yeah, it would be much nicer to be able to just drop down into the code here and remove all the extra stuff in my derived class beyond the customisations to the services and configuration.

Reply:
What's the ideal code you would write?

Reply:
I'll have a think this morning and come up with some pseudo-code.

Reply:
Without trying to think about how it works internally today too much and what I do to get this working in .NET 5 I came up with something like this for what's been changed here to not assume `TestServer` too much.

With the changes in this PR it would make the existing methods to get an HttpClient pointing at the server "just work" removing the need for a custom one that knows to point to the real server not the test server, and also remove the need to create and maintain its own host/server.

```csharp
using System.Security.Cryptography.X509Certificates;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Hosting.Server;
using Microsoft.AspNetCore.Hosting.Server.Features;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace TodoApp;

public sealed class HttpServerFixture : WebApplicationFactory<MyEntryPoint>
{
    public string ServerAddress => ClientOptions.BaseAddress.ToString();

    protected override void ConfigureWebHost(IWebHostBuilder builder)
    {
        base.ConfigureWebHost(builder);

        builder.ConfigureKestrel(
            serverOptions => serverOptions.ConfigureHttpsDefaults(
                httpsOptions => httpsOptions.ServerCertificate = new X509Certificate2("localhost-dev.pfx", "Pa55w0rd!")));

        // Configure the server address for the server to
        // listen on for HTTPS requests on a dynamic port.
        builder.UseUrls("https://127.0.0.1:0");

        // Do something here (if needed?) that would add another IServer implementation to the service collection for a
        // real server so that the call to UseTestServer() here effectively gets no-op'd. The tests would then just not
        // use the ITestServer Server property in the base class hierarchy.
        // https://github.com/dotnet/aspnetcore/blob/aa6c3d27eabfd781bae8d5157d50e1b8eaaf96f0/src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs#L202
    }

    protected override IHost CreateHost(IHostBuilder builder)
    {
        // This assumes the base class has done something that means
        // the host has ended up being a real server, rather than the test server.
        var host = base.CreateHost(builder);

        // Extra the selected dynamic port out of the server and assign it
        // onto the client options for convenience so it "just works" as otherwise
        // it'll be the default http://localhost URL, which won't route to the server.
        var server = host.Services.GetRequiredService<IServer>();
        var addresses = server.Features.Get<IServerAddressesFeature>();

        ClientOptions.BaseAddress = addresses!.Addresses
            .Select((p) => new Uri(p))
            .Last();

        return host;
    }
}
```

I might have a hack around in my test app to see how close to that I could get today without anything being changed here.

Reply:
I think this would be the minimal code needed for my use case with the changes from this PR. Without it fails in `EnsureStarted()` due to the `IServer` not being of type `TestServer` (`System.InvalidCastException : Unable to cast object of type 'Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl' to type 'Microsoft.AspNetCore.TestHost.TestServer'.)`.

```csharp
using System.Security.Cryptography.X509Certificates;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Hosting.Server;
using Microsoft.AspNetCore.Hosting.Server.Features;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace MyApp;

public sealed class HttpServerFixture : WebApplicationFactory<MyEntrypoint>, IAsyncLifetime
{
    public string ServerAddress => ClientOptions.BaseAddress.ToString();

    Task IAsyncLifetime.InitializeAsync()
    {
        using (CreateDefaultClient()) // Force the server to start now
        return Task.CompletedTask;
    }

    Task IAsyncLifetime.DisposeAsync() => Task.CompletedTask;

    protected override void ConfigureWebHost(IWebHostBuilder builder)
    {
        base.ConfigureWebHost(builder);

        builder.ConfigureKestrel(
            serverOptions => serverOptions.ConfigureHttpsDefaults(
                httpsOptions => httpsOptions.ServerCertificate = new X509Certificate2("localhost-dev.pfx", "Pa55w0rd!")));

        // Configure the server address for the server to
        // listen on for HTTPS requests on a dynamic port.
        builder.UseUrls("https://127.0.0.1:0");
    }

    protected override IHost CreateHost(IHostBuilder builder)
    {
        // Use Kestrel instead of TestServer
        builder.ConfigureWebHost((p) => p.UseKestrel());

        var host = base.CreateHost(builder);

        // Extract the selected dynamic port out of the server and assign it
        // onto the client options for convenience so it "just works" as otherwise
        // it'll be the default http://localhost URL, which won't route to the server.
        var server = host.Services.GetRequiredService<IServer>();
        var addresses = server.Features.Get<IServerAddressesFeature>();

        ClientOptions.BaseAddress = addresses!.Addresses
            .Select((p) => new Uri(p))
            .Last();

        return host;
    }
}
```